zkevm-circuits/src/blake2b/circuit/src/lib.rs

```diff
@@ -0,0 +1,1855 @@
+    of the height of a circuit instance is ceil(log2(max((q + 2) * R + (n - 1) * (T + 8), 65536) + u)), where u is the amount of the unusable rows. 
+    
+    For the gates of the InitialGate, RoundGate and FialGate classes the little-endian representations of the used or computed 64-bit values are  
+    taken from or assigned to the specified input byte-column chunks of height 8. For storing the amounts of rounds left to be performed and the  
+    random linear combination hashes the specified unconstrained column cells are used. The 10 binary flags for both the states are stored in the  
+    form of the specified combined selectors, which control the gates computing the m permuted for the next round. The value of f is taken from  
+    the specified bit cell. Some specified cells are used to store the intermediate results of the computation performed by a gate.
+
+    The value of m permuted for the next round is computed by a RoundGate class gate from the value of m permuted for the current round. The new 
+    value is the result of applying a certain index permutation to the current one. An index permutation is an element of the group of permutations 
+    of the set {0, 1, ..., n - 1}, where the n is the array's length. Such a permutation can described by the n-element array, the i-th entry of 
+    which contains the value replacing i. The index permutation described by the array a is designated here as per(a). If an array is considered as 
+    the table with the rows of elements and their indexes, then applying an index permutation to the array means applying this permutation to the 
+    content of second row. So the value of m permuted for the (i + 1) round is the result of applying the index permutation P(i) = p(i + 1) * p'(i) 
+    to the value of m permuted for the i-th round, where p(j) is the index permutation applied to the value m to permute it for the j-th round, 
+    p'(j) is the inverse permutation for p(j) and p(j) * p(k) is the permutation composition, i.e. the permutation, which is equivalent to applying 
+    p(k) and then p(j). Since the value m permuted for the i-th round is (m[SIGMA[i % 10][0]], m[SIGMA[i % 10][1]], ..., m[SIGMA[i % 10][15]]), 
+    the value of SIGMA[i % 10][j] is replaced with j by p(i). On the other hand, per(SIGMA[i % 10]) replaces j with SIGMA[i % 10][j], therefore, 
+    p(i) is per'(SIGMA[i % 10]). So P(i) = per'(SIGMA[(i + 1) % 10]) * per(SIGMA[i % 10]). The permutations computed by this formula are used by 
+    RoundGate class gate to compute the values of m permuted for the next rounds without the need to access the value of m. 
+*/
+
+use std::{ marker::PhantomData, array, convert::TryInto };
+use halo2_proofs::{ arithmetic::FieldExt, circuit::*, plonk::*, poly::Rotation };
+
+// Blake2 initialization vector
+const IV: [u64; 8] = [0x6a09e667f3bcc908, 0xbb67ae8584caa73b, 0x3c6ef372fe94f82b, 0xa54ff53a5f1d36f1, 
```

Comment:
I do not like create a lot of separate files, since it inconvenient to switch between a lot of files in the process of studying code as some integral system. As for constants, the only exception is the case of very long definitions.